pkgs/native_assets_builder/lib/src/build_runner/build_runner.dart

```diff
@@ -193,25 +209,77 @@ class NativeAssetsBuildRunner {
       // Ensure we'll never read outdated build results.
       await buildOutputFile.delete();
     }
-    await runProcess(
+    final arguments = [
+      '--packages=${packageConfigUri.toFilePath()}',
+      buildDotDart.toFilePath(),
+      '--config=${configFile.toFilePath()}',
+    ];
+    final result = await runProcess(
       workingDirectory: workingDirectory,
       executable: dartExecutable,
-      arguments: [
-        '--packages=${packageConfigUri.toFilePath()}',
-        buildDotDart.toFilePath(),
-        '--config=${configFile.toFilePath()}',
-      ],
+      arguments: arguments,
       logger: logger,
       includeParentEnvironment: includeParentEnvironment,
-      expectedExitCode: 0,
-      throwOnUnexpectedExitCode: true,
     );
-    final buildOutput = await BuildOutput.readFromFile(outDir: outDir);
-    final assets = buildOutput?.assets ?? [];
-    validateAssetsPackage(assets, config.packageName);
-    final dependencies = buildOutput?.dependencies.dependencies ?? [];
-    final metadata = dryRun ? null : buildOutput?.metadata;
-    return (assets, dependencies, metadata);
+    var success = true;
+    if (result.exitCode != 0) {
+      final printWorkingDir = workingDirectory != Directory.current.uri;
+      final commandString = [
+        if (printWorkingDir) '(cd ${workingDirectory.toFilePath()};',
+        dartExecutable.toFilePath(),
+        ...arguments.map((a) => a.contains(' ') ? "'$a'" : a),
+        if (printWorkingDir) ')',
+      ].join(' ');
+      logger.severe(
+        '''
+Building native assets for package:${config.packageName} failed.
+build.dart returned with exit code: ${result.exitCode}.
+To reproduce run:
+$commandString
+stderr:
+${result.stderr}
+stdout:
+${result.stdout}
+        ''',
+      );
+      success = false;
+    }
+
+    try {
+      final buildOutput = await BuildOutput.readFromFile(outDir: outDir);
+      final assets = buildOutput?.assets ?? [];
+      success &= validateAssetsPackage(assets, config.packageName);
+      final dependencies = buildOutput?.dependencies.dependencies ?? [];
+      final metadata = dryRun ? null : buildOutput?.metadata;
+      return (assets, dependencies, metadata, success);
+    } on FormatException catch (e) {
+      logger.severe('''
+Building native assets for package:${config.packageName} failed.
+build_output.yaml contained a format error.
+${e.message}
+        ''');
+      success = false;
+      return (<Asset>[], <Uri>[], Metadata({}), false);
+      // TODO(https://github.com/dart-lang/native/issues/109): Stop throwing
+      // type errors in native_assets_cli, release a new version of that package
+      // and then remove this.
+      // ignore: avoid_catching_errors
+    } on TypeError {
```

Comment:
`TypeError`s are usually incorrect programs. Does the yaml logic use that type for exceptions? Maybe it can rather throw a `YamlParseError` or something like that?

Reply:
https://github.com/dart-lang/native/issues/109 tracks this issue. (The issue is in a dependency. Since we have published dependencies, fixing it in the same PR doesn't help, because we'd need to merge first and then publish and then merge this package.)